Type inference free comprehensions

carnaval opened the pull request:
The first part of this is moving lowering of comprehensions to a julia macro. I think it's a good thing in itself.

The second part is to make comprehension have the same semantic as map that is :
- remove type_goto and static_typeof
- use typejoin to compute the array element type
- be optimistic about the first element type

This implies that we get `Vector{Union()}` for 0-length input, as was discussed somewhere else.

The cool things :
- the array type is the same in the repl & in functions, it only depends on the computation not inference. It makes this possible :

``` julia
julia> [rand() > 0.01 ? 1 : "a" for i=1:4, j=1:4]
4x4 Array{Int64,2}:
 1  1  1  1
 1  1  1  1
 1  1  1  1
 1  1  1  1

julia> [rand() > 0.01 ? 1 : "a" for i=1:4, j=1:4]
4x4 Array{Any,2}:
 1  1  1  1   
 1  1  1   "a"
 1  1  1  1   
 1  1  1   "a"
```
- array type is still inferred correctly, with the change that now the result is `Union(Array{Union()},Array{T})`, but it's ok when you index in it as demonstrated by :

``` julia
julia> f() = (A = [1.0*i for i=1:4]; A[1])
f (generic function with 1 method)
julia> Base.return_types(f,())[1]
Float64
```
- in cases where inference gets it right, this should be close to zero cost since the pessimistic codepath gets optimized away. Removing the `Vector{Union()}` after one index call is a bit harder since in general it requires backward dataflow analysis. I didn't check and the macro may be emitting ugly things for now so it may need a bit more work.

The code of the macro is ugly. I've also not yet gone through the frontend to remove dead comprehension lowering code. The dict comprehension case should probably also be written. Sysimg code was also a bit butchered in the process to get it work, we would need to go back and make it right. Thanks to @jakebolewski who was easily baited into doing this.

All in all, not close to done but enough for me to get a type_goto free base for now.

carnaval committed:
Type inference free comprehensions

- Move lowering of the comprehension to a julia macro
- Use typejoin to compute the array/dict type while
  being optimistic about the first element
- Fix comprehension colon logic
- Make tests pass

blakejohnson commented on base/expr.jl:
```diff
@@ -39,7 +37,7 @@ copy(s::SymbolNode) = SymbolNode(s.name, s.typ)
 # copy parts of an AST that the compiler mutates
 astcopy(x::Union(SymbolNode,Expr)) = copy(x)
-astcopy(x::Array{Any,1}) = Any[astcopy(a) for a in x]
+astcopy(x::Array{Any,1}) = map(a->astcopy(a),x)
```
maybe just `map(astcopy, x)`?

carnaval replied:
sure. all of this is a mess and I'll go back to sort it out.

carnaval commented:
Added the dict stuff (works the same way) and cleaned up the dead code from inference & frontend (this PR has a negative line count total).

The generated code for the untyped version of `[X for Y1 in Z1, Y2 in Z2, ...]` looks like :
```julia
index = 1
len = length(Z1)*...
if len == 0
   return Array(Union(), 0)
end
s1 = start(Z1)
Y1,s1 = next(Z1)
s2 = start(Z2)
Y2,s2 = next(Z2)
...
v = X
T = typeof(v)
result = Array(T, len)
@goto inner_loop
while !done(Z1,s1)
   Y1,s1 = next(Z1,s1)
   while !done(Z2,s2)
      Y2,s2 = next(Z2,s2)
      ...
      v = X
      S = typeof(X)
      if !(S <: T)
         T = typejoin(T,S)
         result_next = Array(T, len)
         copy!(result_next, 1, result, 1, index-1)
         result = result_next
      end
      @label inner_loop
      result[index] = v
      index += 1
   end
end
```

carnaval commented:
This now supports colon in comprehension correctly. The current master one is buggy for more than one colon. This now works :
```julia
[[1 2; 3 4] for :, :] == [1 2; 3 4]
```
and any combination of typed/untyped comprehension with colon/non-colon indices also should. The rule being :

Any colon will be substituted by an implicit index into the result of the comprehension expression. The size of this slice is determined by the size of the corresponding slice in the first result.

There is no error checking if you input an expression which changes size (same as on master), also the expression will be recomputed for each iteration of course. This feature seems very underused, but now it should work correctly :-)

carnaval committed:
temp things for gf

carnaval committed:
fix copy(::Array{⊥})

carnaval commented:
Well my job here is "done". Tests pass with some slight changes here and there. I'm not advocating for merge since I'm sure there are some perf regression everywhere due to the type instability. It is useful to me as is so if no one is interested I'm just gonna leave the PR hanging here for now.

JeffBezanson commented on base/essentials.jl:
```diff
@@ -135,7 +135,13 @@ function length_checked_equal(args...)
 end
 
-map(f::Function, a::Array{Any,1}) = Any[ f(a[i]) for i=1:length(a) ]
+function map(f::Function, a::Array{Any,1})
```
Is this rewrite necessary? If so, why?

jakebolewski replied:
The comprehension macro is not defined yet, so if I remember correctly this was an issue during bootstrap (one of the functions in the macro indirectly called this method, I don't know if that is still the case).

JeffBezanson replied:
Ah, of course.

JeffBezanson commented:
This is really cool! Awesome work. I think we can seriously consider merging this. Although it would be nice to *also* base comprehensions on `map`, but that's definitely work for a future release.

tkelman commented on base/inference.jl:
```diff
@@ -2,7 +2,7 @@
 # parameters limiting potentially-infinite types
 const MAX_TYPEUNION_LEN = 3
-const MAX_TYPE_DEPTH = 4
+const MAX_TYPE_DEPTH = 6
```
sneaky... ref #10230

kmsquire commented on base/array.jl:
```diff
@@ -101,6 +101,10 @@ end
 
+function copy(a::Array{Union()})
+    a
+end
```
maybe?
```julia
copy(a:Array{Union()}) = a
```

kmsquire commented on base/operators.jl:
```diff
@@ -351,6 +351,24 @@ for f in (:+, :-)
+
+## non-linear operations on ranges and fallbacks for non-real numbers ##
+
+.+(x::Number, r::Range) = [ x+y for y=r ]
```
Personal opinion: for here, and below, I prefer
```julia
[ x+y for y in r ]
```

carnaval replied:
I just took those verbatim from the range file (they need to be moved around for dependency stuff). I do agree that in is more clear for non-range args in general.

StefanKarpinski replied:
Agreed. The range definition probably predates that syntax.

StefanKarpinski commented:
Is there anything preventing us from merging this?

StefanKarpinski commented:
Aside from the need to rebase due to some (hopefully) minor conflicts?

yuyichao commented:
I think @carnaval was worrying about type stability caused by type instability caused by `Array{Union()}`. I'm also wondering what will happen when there are many `Array{Union()}` passing around....

StefanKarpinski commented:
@JeffBezanson, any thoughts?

yuyichao commented:
And here's @carnaval's comment https://github.com/JuliaLang/julia/issues/7258#issuecomment-110943938.

My concern is largely based on https://github.com/JuliaLang/julia/issues/11658 (which can be fixed in general but probably require more polishing than what we want to wait for 0.4)

StefanKarpinski commented:
Ok, yeah that's too bad. Maybe this belongs in Arraymageddon anyway.

vtjnash commented:
anything from here still useful?
